plugins/module_utils/wdc_redfish_utils.py

```diff
@@ -0,0 +1,542 @@
+    def add_custom_dns(self, domain, ip):
+        """Add a DNS entry to the cache.
+
+        :param str domain: The domain of the DNS entry
+        :param str ip: The IP address to store
+        """
+        # See documentation for socket.getaddrinfo for parameter info
+        if self.is_ipv4(ip):
+            value = (socket.AF_INET,
+                     socket.SOCK_STREAM,
+                     socket.IPPROTO_TCP,
+                     '',
+                     (ip, None))  # Set port to None - will be replaced with requested port
+        else:  # ipv6
+            value = (socket.AF_INET6, 0, 0, '', (ip, None, 0, 0))
+        DnsCacheBypass.dns_cache[domain] = [value]
+
+    @staticmethod
+    def is_ipv4(ip):
+        """Is the given IP address an IPv4 address?
+
+        :param str ip: The IP address to check
+        :rtype: bool
+        """
+        if IPADDRESS_AVAILABLE:
+            return ipaddress.ip_address(ip).version == 4
+        else:
+            # Quick and dirty check if we don't have ipaddress module available
+            return ":" not in ip
+
+
+class WdcRedfishUtils(RedfishUtils):
+    """Extension to RedfishUtils to support WDC enclosures."""
+    # Status codes returned by WDC FW Update Status
+    UPDATE_STATUS_CODE_READY_FOR_FW_UPDATE = 0
+    UPDATE_STATUS_CODE_FW_UPDATE_IN_PROGRESS = 1
+    UPDATE_STATUS_CODE_FW_UPDATE_COMPLETED_WAITING_FOR_ACTIVATION = 2
+    UPDATE_STATUS_CODE_FW_UPDATE_FAILED = 3
+
+    # Status messages returned by WDC FW Update Status
+    UPDATE_STATUS_MESSAGE_READY_FOR_FW_UDPATE = "Ready for FW update"
+    UDPATE_STATUS_MESSAGE_FW_UPDATE_IN_PROGRESS = "FW update in progress"
+    UPDATE_STATUS_MESSAGE_FW_UPDATE_COMPLETED_WAITING_FOR_ACTIVATION = "FW update completed. Waiting for activation."
+    UPDATE_STATUS_MESSAGE_FW_UPDATE_FAILED = "FW update failed."
+
+    def __init__(self,
+                 creds,
+                 root_uris,
+                 timeout,
+                 module,
+                 resource_id,
+                 data_modification):
+        super(WdcRedfishUtils, self).__init__(creds=creds,
+                                              root_uri=root_uris[0],
+                                              timeout=timeout,
+                                              module=module,
+                                              resource_id=resource_id,
+                                              data_modification=data_modification)
+        # Update the root URI if we cannot perform a Redfish GET to the first one
+        self._set_root_uri(root_uris)
+
+    def _set_root_uri(self, root_uris):
+        """Set the root URI from a list of options.
+
+        If the current root URI is good, just keep it.  Else cycle through our options until we find a good one.
+        A URI is considered good if we can GET uri/redfish/v1.
+        """
+        for root_uri in root_uris:
+            uri = root_uri + "/redfish/v1"
+            with DnsCacheBypass(root_uri):
```

Comment:
Why is `DnsCacheBypass` used here - and apparently for most (all?) other requests (I haven't checked all)?

Reply:
I have the same question; I'd hate to set the precedent of bypassing DNS caches on every request. If the DNS cache really needs to be refreshed constantly, it sounds like a networking issue to me where the device needs a short life for the hostname mapping.

Reply:
Due to a quirk in the storage controller, the controller has a tendency to lose its IP address when it is reset, which happens as part of the firmware update process.  When this happens, it takes a while before the client's DNS cache entry is updated.  Therefore, when we supply a hostname in the URL for subsequent Redfish calls, the old IP address will be returned from the DNS cache, and the request is therefore sent to the wrong IP address.  The DNSCacheBypass ensures that we look up the IP address from the DNS server, to avoid the problem.

Reply:
I may be able to work around this by requiring the client to manage the DNS issue, since the UpdateAndActivate command doesn't actually issue subsequent commands after doing the FWActivate.  I will have to test out some alternatives.

Reply:
I will try to get a resolution to this by Monday or shortly thereafter.  Thanks both of you for the feedback!

Reply:
Assuming this is in an environment with a DHCP server, I would expect the client identifier option (option 61) in the DHCP request made from the controller would have static information so that it gets the same IP address on the next reset. At least from working with similar types of devices that will request a new address on each reset, the assignment is stable across resets.

Reply:
Thank you both for the feedback.  I have removed DNSCacheBypass.  It is now the responsibility of the caller to deal with stale DNS entries.